contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,728 @@
+    
+    /**
+     * Runs all the necessary approval functions required for a given ERC20 token.
+     * This function can be called when a new token is added to a set token during a 
+     * rebalance.
+     *
+     * @param _token    Address of the token which needs approval
+     */
+    function approveToken(IERC20 _token) public {
+        _token.approve(address(uniRouter), MAX_UINT96);
+        _token.approve(address(sushiRouter), MAX_UINT96);
+        _token.approve(address(basicIssuanceModule), MAX_UINT96);
+    }
+
+    /* ============ External Functions ============ */
+    
+    /**
+     * Runs all the necessary approval functions required for a list of ERC20 tokens.
+     *
+     * @param _tokens    Addresses of the tokens which need approval
+     */
+    function approveTokens(IERC20[] calldata _tokens) external {
+        for(uint256 i = 0; i < _tokens.length; i++)
+            approveToken(_tokens[i]);
+    }
+
+    /**
+     * Runs all the necessary approval functions required before issuing
+     * or redeeming a set token. This function need to be called only once before the first time
+     * this smart contract is used on any particular set token.
+     *
+     * @param _setToken    Address of the set token being initialized
+     */
+    function approveSetToken(ISetToken _setToken) external {
+        ISetToken.Position[] memory positions = _setToken.getPositions();
+        for (uint256 i = 0; i < positions.length; i++)
+            approveToken(IERC20(positions[i].component));
+    }
+
+    /**
+     * Issues set tokens for an exact amount of input token. 
+     * Acquires set token components at the best price accross uniswap and sushiswap.
+     * Uses the acquired components to issue the set tokens.
+     *
+     * @param _setToken         Address of the set token being issued
+     * @param _inputToken       Address of input token
+     * @param _amountInput      Amount of the input token / ether to spend
+     * @param _minSetReceive    Minimum amount of set tokens to receive
+     */
+    function issueSetForExactToken(
+        ISetToken _setToken,
+        IERC20 _inputToken,
+        uint256 _amountInput,
+        uint256 _minSetReceive
+    )
+        external
+        nonReentrant
+    {   
+        _inputToken.transferFrom(msg.sender, address(this), _amountInput);
+        
+        if(address(_inputToken) != WETH) {      // swap inputToken to WETH
+            (, Exchange exchange) = _getMaxTokenForExactToken(_amountInput, address(_inputToken), WETH);
+            IERC20(_inputToken).approve(address(_getRouter(exchange)), _amountInput);
+            _swapExactTokensForTokens(exchange, address(_inputToken), WETH, _amountInput);
+        }
+            
+        uint256 setTokenAmount = _issueSetForExactWETH(_setToken, _minSetReceive);     // issue set token
+        
+        emit ExchangeIssue(msg.sender, address(_setToken), address(_inputToken), _amountInput, setTokenAmount);
```

Comment:
See note in the events. It would be great if this was the case across the board.